_posts/en/newsletters/2024-05-08-newsletter.md

```diff
@@ -0,0 +1,261 @@
+  check that will commit to the transaction and a lamport signature that
+  commits to the actual size of the ECDSA signature.  For example:
+
+  ```text
+  OP_DUP <pubkey> OP_CHECKSIG OP_SIZE <size> OP_EQUAL
+  OP_IF
+    # Size is equal to <size> bytes
+    OP_SHA256 <digest_x> OP_CHECKEQUALVERIFY
+  OP_ELSE
+    # Size is greater than or less than <size> bytes
+    OP_SHA256 <digest_y> OP_CHECKEQUALVERIFY
+  OP_ENDIF
+  ```
+
+  To satisfy this script fragment, the spender provides an ECDSA
+  signature.  The signature is duplicated and validated; the script
+  fails if it's not a valid signature.  In a post-quantum world, an
+  attacker may be able to pass this test, allowing validation to
+  continue.  The size of the duplicate signature is measured.  If it's
+  equal to `<size>` bytes (e.g., the most common size), the spender must
+  reveal the preimage for `<digest_x>`; otherwise (in the 1/256 case),
+  they must reveal the preimage for `<digest_y>`.  If a valid preimage
+  for the actual signature size isn't revealed, the script fails.
+
+  Now even if ECDSA is completely broken, it's not possible for an
+  attacker to spend the bitcoins unless they also know the lamport
+  private key.  By itself, this isn't very exciting: P2SH and P2WSH
+  already have this fundamental property when their script preimages are
+  kept secret.  However, after the lamport signature is published, an
+  attacker who wants to reuse it with a forged ECDSA signature will have
+  to ensure the ECDSA signature is the same length as the original ECDSA
+  signature.  This may require the attacker grind the signature,
+  performing extra operations that an honest user wouldn't need to
+  perform.
+
+  The amount of grinding an attacker would need to perform can be
+  increased exponentially by including additional pairs of ECDSA and
+  lamport signatures.  Unfortunately, because ECDSA signatures only
+  naturally vary in byte size one time out of every 256, the
+  straightforward way to do this would require a very large number of
+  signatures to be included in order to obtain practical security.
+  Heilman [describes][] a much more efficient mechanism.  That mechanism
+  still exceeds the consensus limits for P2SH <!-- 520 byte redeem
+  script size --> but we think it might just barely work with P2WSH's
+  higher limits<!-- 10,000 byte max script size -->.
+
+  Additionally, an individual attacker with a fast quantum computer or a
+  sufficiently powerful classical computer could discover a short ECDSA
+  nonce that would allow them to easily steal from anyone who didn't
+  anticipate a nonce that short.  The minimal size of a nonce is known,
+  so the attack is avoidable---however, the private form of that nonce
+  is not known, so anyone trying to avoid this attack would be unable to
+  spend their own bitcoins until a fast quantum computer was invented.
+
+  This lamport signature verification is practically similar to the
```

Comment:
```suggestion
  This lamport signature verification is similar to the
```

Reply:
Leaving as-is.  I want to convey that both accomplish the same thing but that they differ in how they accomplish it.  I also want to leave some wiggle room in case the differences are larger than I expect from my initial reading of the idea.  :-)